File: design_pattern/factory.ts
// ============================================================
// Factory Pattern — 工厂函数创建对象，调用方不感知具体类型
// 对比 Rust: 03_factory.rs
// 运行: npx ts-node 03_factory.ts
// ============================================================

interface Logger {
  log(level: string, msg: string): void;
}

class ConsoleLogger implements Logger {
  log(level: string, msg: string) {
    console.log(`[Console][${level}] ${msg}`);
  }
}

class FileLogger implements Logger {
  constructor(private path: string) {}
  log(level: string, msg: string) {
    console.log(`[File:${this.path}][${level}] ${msg}`);
  }
}

class JsonLogger implements Logger {
  log(level: string, msg: string) {
    console.log(JSON.stringify({ level, msg }));
  }
}

function createLogger(kind: string): Logger {
  switch (kind) {
    case "file": return new FileLogger("app.log");
    case "json": return new JsonLogger();
    default:     return new ConsoleLogger();
  }
}

// --- main ---
console.log("=== Factory Pattern ===");

for (const kind of ["console", "file", "json"]) {
  console.log(`\n[${kind}]`);
  const logger = createLogger(kind);
  logger.log("INFO", "server started");
  logger.log("ERROR", "connection refused");
}

File: design_pattern/factory.rs
// ============================================================
// Factory Pattern — 工厂函数创建对象，调用方不感知具体类型
// 对比 TS: 03_factory.ts
// 运行: cargo run --bin factory
// ============================================================

trait Logger {
    fn log(&self, level: &str, msg: &str);
}

struct ConsoleLogger;

struct FileLogger {
    path: String,
}

struct JsonLogger;

impl Logger for ConsoleLogger {
    fn log(&self, level: &str, msg: &str) {
        println!("[Console][{}] {}", level, msg);
    }
}

impl Logger for FileLogger {
    fn log(&self, level: &str, msg: &str) {
        println!("[File:{}][{}] {}", self.path, level, msg);
    }
}

impl Logger for JsonLogger {
    fn log(&self, level: &str, msg: &str) {
        println!(r#"{{"level":"{}","msg":"{}"}}"#, level, msg);
    }
}

fn create_logger(kind: &str) -> Box<dyn Logger> {
    match kind {
        "file" => Box::new(FileLogger { path: "app.log".into() }),
        "json" => Box::new(JsonLogger),
        _      => Box::new(ConsoleLogger),
    }
}

fn main() {
    println!("=== Factory Pattern ===");

    for kind in ["console", "file", "json"] {
        println!("\n[{}]", kind);
        let logger = create_logger(kind);
        logger.log("INFO", "server started");
        logger.log("ERROR", "connection refused");
    }
}

// Rust 关键差异：
// - Box<dyn Logger> 是运行时多态（堆分配），等价于 TS 接口
// - 若编译期类型已知，可用泛型 impl Logger，零堆分配
